test/test_upload_processor.py

```diff
@@ -0,0 +1,155 @@
+""" This file tests the functions defined in upload_processor """
+
+import unittest
+import subprocess
+import os
+from typing import List
+import hailtop.batch as hb
+from joint_calling.upload_processor import batch_move_files
+
+
+def validate_move(upload_prefix: str, main_prefix: str, sample: str) -> bool:
+    """Checks that a given file exists at the location in
+    the main bucket and no longer exists in the upload bucket.
+    Returns True if this is the case and False otherwise.
+    """
+
+    upload_path = os.path.join('gs://', upload_prefix, sample)
+    main_path = os.path.join('gs://', main_prefix, sample)
+
+    exists_main = subprocess.run(['gsutil', '-q', 'stat', main_path], check=False)
+    exists_upload = subprocess.run(['gsutil', '-q', 'stat', upload_path], check=False)
+
+    # Exists at destination and not at source
+    if exists_upload.returncode == 1 and exists_main.returncode == 0:
+        return True
+
+    return False
+
+
+def cleanup(bucket: str, samples: List[str]):
+    """Clean up function that deletes remaining files after
+    test run"""
+
+    for sample in samples:
+        full_path = os.path.join('gs://', bucket, sample)
+        subprocess.run(['gsutil', 'rm', full_path], check=False)
+
+
+def upload_files(files: List[str], upload_prefix: str):
+    """A function to mimic file upload. Takes a list of
+    file names, creates these files, then moved them into
+    the upload bucket used for further testing."""
+
+    for f in files:
+        subprocess.run(['touch', f], check=True)
+        full_path = os.path.join('gs://', upload_prefix, f)
+        subprocess.run(['gsutil', 'mv', f, full_path], check=True)
+
+
+class TestUploadProcessor(unittest.TestCase):
+    """Test cases for the upload processor"""
+
+    def setUp(self):
+        """ Initialises standard variables for testing """
+        self.upload_prefix = 'cpg-fewgenomes-temporary/vivian-test/test-upload'
+        self.main_prefix = 'cpg-fewgenomes-temporary/vivian-test/test-main'
+        self.docker_image = os.environ.get('DOCKER_IMAGE')
+        self.key = os.environ.get('GSA_KEY')
+
+    def test_batch_move_standard(self):
+        """ Testing standard case of moving a list of files with valid inputs"""
+        sample_list = ['Sample5.gVCF', 'Sample6.gVCF', 'Sample7.gVCF']
+        upload_files(sample_list, self.upload_prefix)
+        batch = hb.Batch(name='Test Batch Move Standard')
+        batch_move_files(
+            batch,
+            sample_list,
+            self.upload_prefix,
+            self.main_prefix,
+            self.docker_image,
+            self.key,
+        )
+        batch.run()
```

Comment:
Does this wait for the batch to finish or is it run asynchronously?

Reply:
It waits :)